feat: add lz bridges

souradeep-das opened the pull request:
:clipboard: https://github.com/bobanetwork/boba/issues/235

## Overview

Adds LayerZero compatible bridges to move $BOBA from Eth to other Alt L1s

## Changes

- EthBridge is meant to be deployed on Eth and custodies $BOBA
- AltL1Bridge mints wrapped $BOBA tokens on the altL1 when it hears about deposits on the EthBridge, AltL1Bridge also burns tokens on the altL1 and sends a message to withdraw the actual funds from the custody on EthBridge
- added depl script and tests

## Testing

```
cd packages/boba/contracts
yarn test
```

souradeep-das committed:
draft lz bridges

souradeep-das committed:
make proxy compatible

souradeep-das committed:
add deploy script

souradeep-das committed:
test: add tests

souradeep-das committed:
add hardhat config

souradeep-das commented on packages/boba/contracts/deploy/020-LzBridges.deploy.ts:
```diff
@@ -0,0 +1,169 @@
+
+import EthBridgeJson from '../artifacts/contracts/lzTokenBridge/EthBridge.sol/EthBridge.json'
+import AltL1BridgeJson from '../artifacts/contracts/lzTokenBridge/AltL1Bridge.sol/AltL1Bridge.json'
+import Lib_ResolvedDelegateProxyJson from '../artifacts/contracts/libraries/Lib_ResolvedDelegateProxy.sol/Lib_ResolvedDelegateProxy.json'
```
should we start shifting to a new proxy contract? opinions?

InoMurko replied:
What are the benefits? Cons?

souradeep-das replied:
the current proxy could still have a very very narrow edge-case of collision during upgrades

InoMurko replied:
Bedrock uses a different approach to contract upgrades, afaik. To keep things consistent - is this really problematic for Lz Bridges? Are you seeing us updating these frequently?

souradeep-das replied:
nah, I don't think we will do a lot of upgrades given the nature of the contract

souradeep-das committed:
style: update comments

souradeep-das commented on packages/boba/contracts/contracts/lzTokenBridge/AltL1Bridge.sol:
```diff
@@ -0,0 +1,145 @@
+            IL2StandardERC20(_l2Token).mint(_to, _amount);
+            emit DepositFinalized(_l1Token, _l2Token, _from, _to, _amount, _data);
+        } else {
+            // Either the L2 token which is being deposited-into disagrees about the correct address
```
can I get some 👀  here

since the bridge is only considered to be used for BOBA, user level errors should not be a major reason for this. we have to determine if we still want to offer a way for refund in this case.

if execution gets into the else block, the first time Lz endpoint tries to relay, the relay message would fail, since it wont provide fees for sending the message. The recourse would be for the user to call retryMessage() and paying fees, with this there will still be a way to recover funds in case of user errors.

This looks to be working as expected. Do we see any edge cases here in terms of security?

wsdt replied:
I think claiming your funds manually as an user when the deposit failed is 100% legitimate. I think having a clear error message and Call-to-Action on the Frontend-Side on what to do now, would make it easy enough. 

One thing we could do probably (but I think it is over-engineered), could be to collect an additional fee for deposits/withdrawals and deposit onto e.g. this contract which is used to pay for that crosschain message. If no funds are left, then the fallback as described in the first paragraph would start. Just wanted to mention it here. :)

InoMurko replied:
@souradeep-das are you cool in reach out to L0 on telegram?

souradeep-das replied:
> could be to collect an additional fee for deposits/withdrawals and deposit onto e.g. this contract which is used to pay for that crosschain message. 

great point, yeah one issue with that could be like you said - we have to take some additional funds from every user- when they don't need this reverse send if they are doing everything right. the main issue, i think, with the contract paying for the reverse send is that loading up the contract is async, and someone can get away with the funds you deposited 😄

souradeep-das replied:
> @souradeep-das are you cool in reach out to L0 on telegram?

sure!

wsdt replied:
100%! just wanted to mention it :)

InoMurko replied:
Let's wait for lz here for a day, and see how it goes from there. Otherwise we'll make a call.

souradeep-das committed:
add init checks on depl

wsdt reviewed (approved):
Looks good to me!

souradeep-das committed:
update lz lib contracts to include gas lookup

souradeep-das committed:
add adapter params

souradeep-das commented:
added in some (mostly) LayerZero future parameters as per their suggestion-
 - added gap to the upgradeable libraries
 - added LayerZero's customAdapterParameter to allow for custom gasLimits for xMessages (useful when the logic exceeds 200k gas)
 - added the option to specify zroTokenAddres - to support paying messaging fees in that denomination
 
can I get re-reviews for the last two commits 🙏

InoMurko commented on packages/boba/contracts/test/contracts/lzTokenBridges/lzBridges.spec.ts:
```diff
@@ -539,4 +562,109 @@ describe('LayerZero Bridges', () => {
     ).to.be.revertedWith('NonblockingLzApp: invalid payload')
   })
+
+  it('only the owner should be able to allow custom adapter params', async function () {
```
haha, I think you went all in on testing this:
owner - set 220000
check if 220000
not owner - set 180000
owner - set 180000
check if 180000
owner - set 200000 (reset)
check if 200000

have I nailed the flow? might be useful to leave it as a comment :-)

souradeep-das replied:
hahaha, precisely - also because 220k and 180k are towards separate contracts 😂 
added some comments 😋

InoMurko replied:
Ahhh, you know I didn't notice that

wsdt commented:
New changes look great! (already approved)
Also love the test Ino pointed out 😂

souradeep-das committed:
add comments

boyuan-chen commented:
Could you please populate the changes to the PR in boba repo?

souradeep-das commented:
done, synced the changes to the other PR
